funding/batch.go

```diff
@@ -0,0 +1,512 @@
+	// verifying the passed PSBT.
+	PsbtFundingVerify([32]byte, *psbt.Packet) error
+
+	// PsbtFundingFinalize looks up a previously registered funding intent
+	// by its pending channel ID and tries to advance the state machine by
+	// finalizing the passed PSBT.
+	PsbtFundingFinalize([32]byte, *psbt.Packet, *wire.MsgTx) error
+
+	// PublishTransaction performs cursory validation (dust checks, etc),
+	// then finally broadcasts the passed transaction to the Bitcoin network.
+	PublishTransaction(*wire.MsgTx, string) error
+
+	// CancelFundingIntent allows a caller to cancel a previously registered
+	// funding intent. If no intent was found, then an error will be
+	// returned.
+	CancelFundingIntent([32]byte) error
+}
+
+// BatchConfig is the configuration for executing a single batch transaction for
+// opening multiple channels atomically.
+type BatchConfig struct {
+	// RequestParser is the function that parses an incoming RPC request
+	// into the internal funding initialization message.
+	RequestParser RequestParser
+
+	// ChannelOpener is the function that kicks off the initial channel open
+	// negotiation with the peer.
+	ChannelOpener ChannelOpener
+
+	// ChannelAbandoner is the function that can abandon a channel in the
+	// local database, graph and arbitrator state.
+	ChannelAbandoner ChannelAbandoner
+
+	// WalletKitServer is an instance of the wallet kit sub server that can
+	// handle PSBT funding and finalization.
+	WalletKitServer WalletKitServer
+
+	// Wallet is an instance of the internal lightning wallet.
+	Wallet Wallet
+
+	// NetParams contains the current bitcoin network parameters.
+	NetParams *chaincfg.Params
+
+	// Quit is the channel that is selected on to recognize if the main
+	// server is shutting down.
+	Quit chan struct{}
+}
+
+// Batcher is a type that can be used to perform an atomic funding of multiple
+// channels within a single on-chain transaction.
+type Batcher struct {
+	cfg *BatchConfig
+
+	channels    []*batchChannel
+	lockedUTXOs []*walletrpc.UtxoLease
+
+	didPublish bool
+
+	wg           sync.WaitGroup
+	negotiateErr error
+}
+
+// NewBatcher returns a new batch channel funding helper.
+func NewBatcher(cfg *BatchConfig) *Batcher {
+	return &Batcher{
+		cfg: cfg,
+	}
+}
+
+// BatchFund starts the atomic batch channel funding process.
+//
+// NOTE: This method should only be called once per instance.
+func (b *Batcher) BatchFund(ctx context.Context,
+	req *lnrpc.BatchOpenChannelRequest) ([]*lnrpc.PendingUpdate, error) {
+
+	label, err := labels.ValidateAPI(req.Label)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse and validate each individual channel.
+	b.channels = make([]*batchChannel, 0, len(req.Channels))
+	for idx, rpcChannel := range req.Channels {
+		var pendingChanID [32]byte
+		if _, err := rand.Read(pendingChanID[:]); err != nil {
+			return nil, fmt.Errorf("error making temp chan ID: %v",
+				err)
+		}
+
+		fundingReq, err := b.cfg.RequestParser(&lnrpc.OpenChannelRequest{
```

Comment:
I would think we can just use the chan request as is given there appear to be no extra fields? Other than the requirement of no-publish for the PSBt state.

Reply:
You mean use `lnrpc.OpenChannelRequest` instead of adding a new `lnrpc.BatchOpenChannel` type?
My goal was to also create a non-confusing API with just the fields per channel that are really needed, minus the deprecated ones.